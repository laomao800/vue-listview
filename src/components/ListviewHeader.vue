<template>
  <div class="lv__header">
    <h1 v-if="headerTitle" class="lv__header-title">{{ headerTitle }}</h1>
    <el-breadcrumb
      v-if="internalNav.length > 0"
      separator="/"
      class="lv__header-breadcrumb"
    >
      <el-breadcrumb-item
        v-for="(item, index) in internalNav"
        :key="index"
        :to="item.to"
      >
        {{ item.text }}
      </el-breadcrumb-item>
    </el-breadcrumb>
  </div>
</template>

<script>
import isPlainObject from 'lodash/isPlainObject'

export default {
  name: 'ListviewHeader',

  inheritAttrs: false,

  props: {
    headerTitle: { type: String, default: '' },
    headerNav: { type: Array, default: () => [] },
  },

  computed: {
    internalNav() {
      const validNav = []
      this.headerNav.forEach((nav) => {
        let text, to
        if (typeof nav === 'string' && !!nav) {
          text = nav
        } else if (isPlainObject(nav)) {
          text = nav.text
          to = nav.to
        }
        if (text) {
          validNav.push({ text, to })
        }
      })
      return validNav
    },
  },
}
</script>

<style lang="less">
.lv__header {
  margin: -10px -10px 10px;
  padding: 8px 12px;
  line-height: 1.5;
  background-color: #fff;
  border-bottom: 1px solid #dcdfe6;

  &:empty {
    display: none;
  }

  .lv__header-title {
    display: inline-block;
    padding: 0 1em 0 0;
    margin: 0 1em 0 0;
    font-size: 16px;
    border-right: 1px solid #dcdfe6;
  }
  .lv__header-breadcrumb {
    display: inline-block;
    vertical-align: middle;
  }
}
</style>
