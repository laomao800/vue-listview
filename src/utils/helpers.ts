import { VNode } from 'vue'
import isPlainObject from 'lodash/isPlainObject'
import isEmpty from 'lodash/isEmpty'
import isFunction from 'lodash/isFunction'
import get from 'lodash/get'
import { FilterField } from '~/types'

/**
 * 判断 node 是否为 Vue 内部的 VNode 类型
 */
export function isVNode(node: any): node is VNode {
  return (
    node !== null &&
    typeof node === 'object' &&
    hasOwn(node, 'componentOptions')
  )
}

/**
 * 根据映射表设置的结构提取数据并返回一个新对象
 *
 * dataMapping({
 *   result: {
 *     results: [1, 2, 3]
 *   }
 * }, {
 *   items: 'result.results'
 * })
 *
 * -> { items: [1, 2, 3] }
 *
 * @param {Object} data 目标数据
 * @param {Object} dataMap 映射表
 */
export function dataMapping(
  data: Record<string, any> = {},
  dataMap: Record<string, any> = {}
) {
  const result: Record<string, any> = {}

  if (isPlainObject(dataMap)) {
    Object.keys(dataMap).forEach((key) => {
      try {
        const dataKey = key.toString()
        const dataValue = get(data, dataMap[key])
        result[dataKey] = dataValue
      } catch (e) {}
    })
  }

  return result
}

/**
 * 判断值是否为搜索栏内合法的值，通过验证的值才可继续作为参数随请求提交
 */
export function isValidFieldValue(val: any): val is FilterField {
  return !(
    val === null ||
    val === undefined ||
    val === '' ||
    ((Array.isArray(val) || isPlainObject(val)) && isEmpty(val))
  )
}

/**
 * 判断一个值是否有效的搜索栏 field 配置项
 */
export function isValidFieldConfig(field: any) {
  return (
    field &&
    (hasOwn(field, 'type') ||
      isFunction(field.render) ||
      isFunction(field) ||
      isVNode(field))
  )
}

export function nodeParents(node: Element, selector: string) {
  const allMatchs = Array.from(document.querySelectorAll(selector))
  if (allMatchs.length === 0) {
    return null
  }
  function find(curNode: Element): Element | null {
    const parentNode = curNode.parentNode as Element | null
    if (parentNode) {
      if (allMatchs.includes(parentNode)) {
        return parentNode
      } else if (parentNode.parentNode) {
        return find(parentNode)
      }
    }
    return null
  }
  return find(node)
}

export function isPromise(obj: any) {
  return (
    !!obj &&
    (typeof obj === 'object' || typeof obj === 'function') &&
    typeof obj.then === 'function'
  )
}

export function hasOwn(obj: Record<string, unknown>, key: string) {
  return Object.prototype.hasOwnProperty.call(obj, key)
}

export function isDef(val: any) {
  return val !== undefined
}

export function hasRenderFn<T>(item: any): item is T {
  return isPlainObject(item) && isFunction(item.render)
}

export function ensurePromise<T>(data: T) {
  return isPromise(data) ? data : Promise.resolve(data)
}

export function noop() {
  // noop
}

export function resolveOptions(
  optionsConfig: any,
  done: (options: any[]) => void
) {
  let optionsPromise = null
  if (optionsConfig) {
    if (Array.isArray(optionsConfig)) {
      optionsPromise = Promise.resolve(optionsConfig)
    } else if (isFunction(optionsConfig)) {
      optionsPromise = ensurePromise(optionsConfig(done))
    } else if (isPromise(optionsConfig)) {
      optionsPromise = optionsConfig
    }
  }
  return optionsPromise
}
