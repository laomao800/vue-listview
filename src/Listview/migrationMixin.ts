import Vue from 'vue'
import get from '@/utils/getValue'
import camelCase from 'lodash/camelCase'
import { warn } from '@/utils/debug'
import { hasOwn } from '@/utils'

const insteadWarn = (type: string, oldName: string, newName: string) =>
  warn(
    `[Migration][${type}]: '${oldName}' is removed, use '${newName}' instead.`
  )

export default Vue.extend({
  mounted() {
    // Props
    const props = get(this, '$vnode.data.attrs', {})
    const migrationProps: Record<string, string> = {
      tableProps: 'contentProps',
      tableEvents: 'contentEvents',
      tableSelection: 'selection',
    }
    Object.entries(props).forEach(([propName, propValue]) => {
      propName = camelCase(propName)
      if (migrationProps[propName]) {
        insteadWarn('Props', propName, migrationProps[propName])
      }
      if (propName === 'filterFields') {
        if (
          Array.isArray(propValue) &&
          propValue.some((field: any) => hasOwn(field, 'get'))
        ) {
          insteadWarn('Props', 'filterFields.get()', 'transformRequestData()')
        }
      }
      if (propName === 'contentMessage') {
        if (hasOwn(propValue as any, 'message')) {
          insteadWarn('Props', 'contentMessage.message', 'contentMessage.text')
        }
      }
    })

    // Slots
    const migrationSlots: Record<string, string> = {
      'prepend-filterbar-submit': 'prepend-submit',
      'append-filterbar-submit': 'append-submit',
      'prepend-filterbar-more': 'prepend-more',
      'append-filterbar-more': 'append-more',
    }
    Object.keys(this.$slots).forEach((name) => {
      if (migrationSlots[name]) {
        insteadWarn('Slots', name, migrationSlots[name])
      }
    })
  },
})
