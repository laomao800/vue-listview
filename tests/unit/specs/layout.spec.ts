import { mount } from '@vue/test-utils'
import { wait } from '../helpers'
import Listview from '@/Listview'
import ListviewHeader from '@/components/ListviewHeader.vue'

describe('layout', () => {
  describe('height', () => {
    it('full height', async () => {
      Object.defineProperty(window, 'innerHeight', {
        writable: true,
        configurable: true,
        value: 800,
      })

      const wrapper = await mount(Listview, {
        propsData: { autoload: false, usePage: false },
      })
      expect(wrapper.element.style.height).toBe('800px')

      Object.defineProperty(window, 'innerHeight', {
        writable: true,
        configurable: true,
        value: 500,
      })
      window.dispatchEvent(new Event('resize'))
      await wait()
      expect(wrapper.element.style.height).toBe('500px')
    })

    it('auto height', async () => {
      const wrapper = await mount(Listview, {
        propsData: { autoload: false, usePage: false, fullHeight: false },
      })
      expect(wrapper.element.style.height).toBe('')
    })

    it('specify height', async () => {
      const wrapper = await mount(Listview, {
        propsData: { autoload: false, usePage: false, height: 500 },
      })
      expect(wrapper.element.style.height).toBe('500px')
    })
  })

  describe('ListviewHeader', () => {
    it('empty', async () => {
      const wrapper = await mount(ListviewHeader)
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('title', () => {
      const headerTitle = 'headerTitle'
      const wrapper = mount(ListviewHeader, {
        propsData: { headerTitle },
      })
      expect(wrapper.text()).toMatch(headerTitle)
    })

    it('nav', async () => {
      const wrapper = await mount(ListviewHeader, {
        propsData: {
          headerNav: ['home', { text: 'list', to: '/list' }],
        },
      })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('title & nav', async () => {
      const wrapper = await mount(ListviewHeader, {
        propsData: {
          headerTitle: 'headerTitle',
          headerNav: [
            { text: 'home', to: '/home' },
            { text: 'list', to: '/list' },
          ],
        },
      })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('contentMessage', () => {
    it('string', () => {
      const contentMessage = 'message text'
      const wrapper = mount(Listview, {
        propsData: { contentMessage },
      })
      const messageBlock = wrapper.find('.lv__message')
      expect(messageBlock.find('.el-icon-warning').exists()).toBe(true)
      expect(messageBlock.find('.lv__message-text').text()).toMatch(
        contentMessage
      )
    })

    it('object', () => {
      const contentMessage = {
        type: 'info',
        text: 'message text',
      }
      const wrapper = mount(Listview, {
        propsData: { contentMessage },
      })
      const messageBlock = wrapper.find('.lv__message')
      expect(messageBlock.find('.el-icon-info').exists()).toBe(true)
      expect(messageBlock.find('.lv__message-text').text()).toMatch(
        contentMessage.text
      )
    })
  })

  describe('row class name', () => {
    it('row string class', async () => {
      const wrapper = await mount(Listview, {
        propsData: {
          requestHandler: () => ({
            result: { items: [{}], total: 1 },
            is_success: true,
          }),
          tableProps: { rowClassName: 'row-view-class' },
        },
      })
      await wait()
      expect(
        wrapper.findAll('.el-table__body-wrapper .el-table__row.row-view-class')
          .length
      ).toBe(1)
    })

    it('row func class', async () => {
      const wrapper = await mount(Listview, {
        propsData: {
          requestHandler: () => ({
            result: { items: [{}], total: 1 },
            is_success: true,
          }),
          tableProps: {
            rowClassName: () => 'row-view-class-fn',
          },
        },
      })
      await wait()
      expect(
        wrapper.findAll(
          '.el-table__body-wrapper .el-table__row.row-view-class-fn'
        ).length
      ).toBe(1)
    })
  })

  describe('listview footer', () => {
    it('pager off', async () => {
      const wrapper = await mount(Listview, { propsData: { usePage: false } })
      expect(wrapper.vm.$el.querySelector('.lv__pager')).toBe(null)
    })

    it('pager on left', async () => {
      const wrapper = await mount(Listview, {
        propsData: { pagePosition: 'left' },
      })
      expect(wrapper.find('.lv__footer-left .lv__pager').exists()).toBe(true)
      expect(wrapper.find('.lv__footer-right .lv__pager').exists()).toBe(false)
    })

    it('pager on right', async () => {
      const wrapper = await mount(Listview, {
        propsData: { pagePosition: 'right' },
      })
      expect(wrapper.find('.lv__footer-left .lv__pager').exists()).toBe(false)
      expect(wrapper.find('.lv__footer-right .lv__pager').exists()).toBe(true)
    })

    it('footer slots', async () => {
      const slots = {
        'footer-left': 'footer left',
        'footer-center': 'footer center',
        'footer-right': 'footer right',
      }
      const wrapper = await mount(Listview, { slots })
      expect(wrapper.find('.lv__footer-left').text()).toBe(slots['footer-left'])
      expect(wrapper.find('.lv__footer-center').text()).toBe(
        slots['footer-center']
      )
      expect(wrapper.find('.lv__footer-right').text()).toBe(
        slots['footer-right']
      )
    })
  })
})
