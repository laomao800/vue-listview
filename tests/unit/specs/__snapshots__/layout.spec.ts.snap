// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListviewHeader empty 1`] = `
<div class="lv__header">
  <!---->
  <!---->
</div>
`;

exports[`ListviewHeader nav 1`] = `
<div class="lv__header">
  <!---->
  <div aria-label="Breadcrumb" role="navigation" class="el-breadcrumb lv__header-breadcrumb"><span class="el-breadcrumb__item"><span role="link" class="el-breadcrumb__inner">
      home
    </span><span role="presentation" class="el-breadcrumb__separator">/</span></span><span class="el-breadcrumb__item" aria-current="page"><span role="link" class="el-breadcrumb__inner is-link">
      list
    </span><span role="presentation" class="el-breadcrumb__separator">/</span></span></div>
</div>
`;

exports[`ListviewHeader title & nav 1`] = `
<div class="lv__header">
  <h1 class="lv__header-title">headerTitle</h1>
  <div aria-label="Breadcrumb" role="navigation" class="el-breadcrumb lv__header-breadcrumb"><span class="el-breadcrumb__item"><span role="link" class="el-breadcrumb__inner is-link">
      home
    </span><span role="presentation" class="el-breadcrumb__separator">/</span></span><span class="el-breadcrumb__item" aria-current="page"><span role="link" class="el-breadcrumb__inner is-link">
      list
    </span><span role="presentation" class="el-breadcrumb__separator">/</span></span></div>
</div>
`;

exports[`layout ListviewHeader empty 1`] = `
<div class="lv__header">
  <!---->
  <!---->
</div>
`;

exports[`layout ListviewHeader nav 1`] = `
<div class="lv__header">
  <!---->
  <div aria-label="Breadcrumb" role="navigation" class="el-breadcrumb lv__header-breadcrumb"><span class="el-breadcrumb__item"><span role="link" class="el-breadcrumb__inner">
      home
    </span><span role="presentation" class="el-breadcrumb__separator">/</span></span><span class="el-breadcrumb__item" aria-current="page"><span role="link" class="el-breadcrumb__inner is-link">
      list
    </span><span role="presentation" class="el-breadcrumb__separator">/</span></span></div>
</div>
`;

exports[`layout ListviewHeader title & nav 1`] = `
<div class="lv__header">
  <h1 class="lv__header-title">headerTitle</h1>
  <div aria-label="Breadcrumb" role="navigation" class="el-breadcrumb lv__header-breadcrumb"><span class="el-breadcrumb__item"><span role="link" class="el-breadcrumb__inner is-link">
      home
    </span><span role="presentation" class="el-breadcrumb__separator">/</span></span><span class="el-breadcrumb__item" aria-current="page"><span role="link" class="el-breadcrumb__inner is-link">
      list
    </span><span role="presentation" class="el-breadcrumb__separator">/</span></span></div>
</div>
`;
